contracts/token/IERC20WithPermit.sol

```diff
@@ -0,0 +1,69 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+
+/// @title  IERC20WithPermit
+/// @notice Burnable ERC20 token with EIP2612 permit functionality. User can
+///         authorize a transfer of their token with a signature conforming
+///         EIP712 standard instead of an on-chain transaction from their
+///         address. Anyone can submit this signature on the user's behalf by
+///         calling the permit function, as specified in EIP2612 standard,
+///         paying gas fees, and possibly performing other actions in the same
+///         transaction.
+interface IERC20WithPermit is IERC20 {
+    /// @notice EIP2612 approval made with secp256k1 signature.
+    ///         Users can authorize a transfer of their tokens with a signature
+    ///         conforming EIP712 standard, rather than an on-chain transaction
+    ///         from their address. Anyone can submit this signature on the
+    ///         user's behalf by calling the permit function, paying gas fees,
+    ///         and possibly performing other actions in the same transaction.
+    /// @dev    The deadline argument can be set to uint(-1) to create permits
+    ///         that effectively never expire.
+    function permit(
+        address owner,
+        address spender,
+        uint256 amount,
+        uint256 deadline,
+        uint8 v,
+        bytes32 r,
+        bytes32 s
+    ) external;
+
+    /// @notice Destroys `amount` tokens from the caller.
+    function burn(uint256 amount) external;
+
+    /// @notice Destroys `amount` of tokens from `account`, deducting the amount
+    ///         from caller's allowance.
+    function burnFrom(address account, uint256 amount) external;
+
+    /// @notice Executes receiveApproval function on spender as specified in
+    ///         IReceiveApproval interface. Approves spender to withdraw from
+    ///         the caller multiple times, up to the value amount. If this
+    ///         function is called again, it overwrites the current allowance
+    ///         with value. Reverts if the approval reverted or if
+    ///         receiveApproval call on the spender reverted.
+    function approveAndCall(
```

Comment:
Our `ERC20WithPermit` implements it. I was looking into extracting the implementation to another contract but then it is getting hairy because that contract needs to assume there is a base ERC20 contract that implements `approve`. And we start implementing a framework then... I think the gain is not worth the complexity.

Reply:
Done in https://github.com/thesis/solidity-contracts/pull/3/commits/5596b66132ca76349036f57a29cb55c95651cc6a.